### ISSUE TITLE
您提出的这个数据增强的方法，是否可以用到其他数据集上

### ISSUE THREAD

**Issue by 1chenchen22** [fistyee/MixPro#6]
作者，您好，想请问一下，您提出的这个数据增强的方法，是否可以用到其他数据集上，
比如我想用到我的 人脸表情识别数据集上，对不平衡类别的数据集做数据增强是可以的吗，将 将MixPro 数据增强方法应用到我的模型，具体要注意哪些吗
感谢您的回答

**Comment by fistyee** [comment]
感谢您的关注，用在其他数据集上也是没问题的，相关代码在utils.py/ def MaskMix里。
这个方法可以和Mixup/CutMix一样在训练时加载。
原版代码的输入尺寸是(224x224)，ViT的patch大小为14x14，MaskMix的Mask交换图像大小56x56。
你可以根据你的需求修改输入参数，使得这个方法适用于你的输入图像尺寸和ViT的patch大小。

**Comment by 1chenchen22** [comment]
您好，我做了两种尝试

-  一、创建了一个 RafDataSet 实例 train_dataset，RafDataSet 类是加载我的raf人脸表情数据集的，然后调用 apply_maskmix 方法对数据集进行增强，然后把增强后的数据集赋值给train_dataset放到模型中做训练。

数据增强方法 apply_maskmix：遍历数据集中的图像和标签，将它们按照标签分组。对于每个标签，计算其对应的增强因子，并根据增强因子计算 alpha 和 mask 数量。我是想让类别数量少的标签类别，做更多的数据增强，但是不知道这样操作对不对
[defaultdict(<class 'int'>, {4: 1982, 3: 4772, 0: 1290, 5: 705, 1: 281, 2: 717, 6: 2524})

Class 5: Enhancement Factor: 0.3787575150300601, Alpha: 1.3787575150300602, Mask Num: 8

Class 4: Enhancement Factor: 1.0, Alpha: 2.0, Mask Num: 15

Class 1: Enhancement Factor: 0.22467156535292807, Alpha: 1.224671565352928, Mask Num: 7

Class 6: Enhancement Factor: 0.0944110443108439, Alpha: 1.094411044310844, Mask Num: 5

Class 2: Enhancement Factor: 0.0, Alpha: 1.0, Mask Num: 5

Class 3: Enhancement Factor: 0.09708305499888666, Alpha: 1.0970830549988866, Mask Num: 5

Class 7: Enhancement Factor: 0.4994433311066578, Alpha: 1.4994433311066577, Mask Num: 9
Whole train set size: 12271
Validation set size: 3068](url)
这里是打印出来的参数，

1. 经过了maskmix训练集的数量不应该变多吗，但是这里还是12271没有变，这是我觉得奇怪的第一个点
2. 我还想问一下您，我如果想处理类别不平衡的数据集，就是让，类别数量少的做更多的数据增强，那这里的alpha和mask-num要这么设置比较好
3. 这样做得到的效果并不好，我训练了40轮
这里是得到的结果，还没有之前好，之前的准确率是不到91%

```

2024-04-12 19:42 INFO [Epoch 40]  Training accuracy:  0.9847. Loss: 0.962
2024-04-12 19:42 INFO [Epoch 40]  Validation accuracy:  0.8207. Loss: 1.807
2024-04-12 19:42 INFO Best_acc:0.8305



class RafDataSet(data.Dataset):
    def __init__(self, raf_path, phase, transform=None):
        self.phase = phase
        self.transform = transform
        self.raf_path = raf_path

        df = pd.read_csv(
            os.path.join(self.raf_path, "EmoLabel/list_patition_label.txt"),
            sep=" ",
            header=None,
            names=["name", "label"],
        )

        if phase == "train":
            self.data = df[df["name"].str.startswith("train")]
        else:
            self.data = df[df["name"].str.startswith("test")]

        file_names = self.data.loc[:, "name"].values
        # 0:Surprise, 1:Fear, 2:Disgust, 3:Happiness, 4:Sadness, 5:Anger, 6:Neutral
        self.label = self.data.loc[:, "label"].values - 1

        self.class_counts = defaultdict(int)  # 存储每个类别的样本数量
        for label in self.label:
            self.class_counts[label] += 1

        self.file_paths = []
        for f in file_names:
            f = f.split(".")[0]
            f = f + "_aligned.jpg"
            path = os.path.join(self.raf_path, "aligned", f)
            self.file_paths.append(path)

        # 计算增强因子
        self.enhancement_factors = {label: (count - min(self.class_counts.values())) / (
                    max(self.class_counts.values()) - min(self.class_counts.values())) for label, count in
                               self.class_counts.items()}

    def __len__(self):
        return len(self.file_paths)

    def __getitem__(self, idx):
        print(self.class_counts)
        path = self.file_paths[idx]
        image = cv2.imread(path)
        label = self.label[idx]
        image = Image.fromarray(image[:, :, ::-1])
        if self.transform is not None:
            image = self.transform(image)
        else:
            image = ToTensor()(image)  # 如果未指定 transform，则使用默认的 ToTensor 转换

        return image, label

    def apply_maskmix(self, scale_, scale, label_smoothing, num_classes):
        grouped_imgs = defaultdict(list)
        for img, target in self:
          
            grouped_imgs[target].append(img)

        processed_images = []
        processed_labels = []

        for targets, img_group in grouped_imgs.items():
            if len(img_group) > 1:
                enhancement_factor = self.enhancement_factors[targets]
                print(enhancement_factor)
                mask_num = calculate_mask_num(enhancement_factor)
                alpha = calculate_alpha(enhancement_factor)  # 将 enhancement_factor 作为参数传递
                # 打印增强参数 alpha 和 mask 数量
                print(
                    f"Class {targets + 1}: Enhancement Factor: {enhancement_factor}, Alpha: {alpha}, Mask Num: {mask_num}")
                img_group_tensor = torch.stack(img_group)
                maskmix_imgs, maskmix_targets, _, _ = MaskMix(img_group_tensor, None, alpha, mask_num, scale_, scale,
                                                              torch.tensor([targets] * len(img_group)),
                                                              None, label_smoothing, num_classes)
                processed_images.extend(maskmix_imgs)
                processed_labels.extend(maskmix_targets)

        return processed_images, processed_labels

def calculate_alpha(enhancement_factor):  # 接受 enhancement_factor 作为参数

    根据每个类别的样本数量差异计算增强参数 alpha

    # 根据增强程度调整 alpha 参数
    alpha = 1.0 + enhancement_factor  # 可根据具体情况调整
    return alpha

def calculate_mask_num(enhancement_factor):

    根据增强程度确定每个类别的 mask 数量

    mask_num = int(5 + 10 * enhancement_factor)  # 可根据具体情况调整
    return mask_num

```

- 第二种尝试
我是在训练循环迭代里面做的maskmix， 得到的效果和之前差不多，90。
这样做得到的训练集是98168，就是数量多了很多，比我之前的
```

2024-04-12 14:15 INFO [Epoch 40]  Training accuracy:  0.9914. Loss: 0.326
2024-04-12 14:15 INFO [Epoch 40]  Validation accuracy:  0.8944. Loss: 1.044
2024-04-12 14:15 INFO Best_acc:0.9009


  for epoch in tqdm(range(1, args.epochs + 1)):
        running_loss = 0.0
        correct_sum = 0
        iter_cnt = 0
        model.train()

        for (imgs, targets) in train_loader:
            iter_cnt += 1
            optimizer.zero_grad()
            imgs = imgs.to(device)
            targets = targets.to(device)
           # print("targets",targets.shape)
            # 清空列表
            maskmix_imgs_list = []
            maskmix_targets_list = []
            # 在每个表情类别内进行 MaskMix
            for label in targets.unique():  # 遍历每个唯一的标签
                maskmix_group = (targets == label).nonzero().squeeze()  # 找到具有相同标签的图像索引
                img_group = imgs[maskmix_group]
                targets_group = targets[maskmix_group]  # 获取具有相同标签的子集的目标张量
                print("targets_group", targets_group.shape)

                # 应用 MaskMix 到具有相同表情的图像上
                maskmix_imgs, maskmix_targets, _, _ = MaskMix(img_group, None, alpha, mask_num, scale_, scale,
                                                                  targets_group, None, label_smoothing,
                                                                  num_classes)

                maskmix_imgs_list.append(maskmix_imgs)
                maskmix_targets_list.append(maskmix_targets.unsqueeze(0))  # 将张量添加到列表时增加一个维度
                # 打印每个张量的形状
                print("maskmix_targets shape:", maskmix_targets.shape)
            # 检查列表是否为空，如果不为空，则进行拼接
            if maskmix_targets_list:
                # 将列表中的张量拼接成一个张量
                maskmix_imgs = torch.cat(maskmix_imgs_list, dim=0)
                maskmix_targets = torch.cat(maskmix_targets_list, dim=0)
                print("maskmix_targets shape:", maskmix_targets.shape)
                print("maskmix_imgs shape:", maskmix_imgs.shape)
                # 在这里执行模型的前向传播和损失计算
            out, feat, heads = model(maskmix_imgs)
            loss = criterion_cls(out, maskmix_targets) + 0.1 * criterion_at(heads) + 1 * criterion_af(feat, maskmix_targets)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()

```

我还没有仔细看您的文章，我也不是很懂，明天我再试试看看，麻烦您能不能也帮我看看我这里的maskmix参数要怎么修改，才会让这个效果变得更好吗，谢谢您，非常感谢。 @fistyee

**Comment by 1chenchen22** [comment]
您好，我想请问一下
1.  就是我如果用了maskmix还需要对数据集做data_transforms的增强操作吗



```
data_transforms = transforms.Compose([
        transforms.Resize((112, 112)),
        transforms.RandomHorizontalFlip(),
        transforms.RandomApply([
            transforms.RandomRotation(5),
            transforms.RandomCrop(112, padding=8)
        ], p=0.2),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.485, 0.456, 0.406],
                             std=[0.229, 0.224, 0.225]),
        transforms.RandomErasing(scale=(0.02, 0.25)),
    ])
    train_dataset = datasets.ImageFolder(f'{args.raf_path}/train',transform = data_transforms)
    # 应用处理方法
    processed_images, processed_labels = apply_maskmix(train_dataset, scale_, scale, label_smoothing, num_classes)
```

2. 我换了另外一种方式，就是输出了混淆矩阵，对混淆矩阵里面类别数量少，并且准确率低的几个类别做操作，下面是我的代码
 ```
  # 定义一个方法来处理目标类别为 1、2、4 和 6 的图像
    def apply_maskmix(train_dataset, scale_, scale, label_smoothing, num_classes):
        enhanced_counts = defaultdict(int)
        grouped_imgs = defaultdict(list)
        processed_images = []
        processed_labels = []

        for img, target in train_dataset:
            if target in [1, 2, 4, 6]:
                grouped_imgs[target].append((img, target))  # 将图片和标签一起存储
            else:
                # 对于不需要增强的类别，直接添加原始样本
                processed_images.append(img)
                processed_labels.append(target)
                enhanced_counts[target] += 1

        for targets, img_group in grouped_imgs.items():
           
            img_group_tensor = torch.stack([(item[0]) for item in img_group])  # 取出图片张量
            maskmix_imgs, maskmix_targets, _, _ = MaskMix(img_group_tensor, None, alpha, mask_num, scale_,
                                                          scale, torch.tensor([targets] * len(img_group)),
                                                          None, label_smoothing, num_classes)
            processed_images.extend(maskmix_imgs.tolist())
            processed_labels.extend(maskmix_targets.tolist())
            # 统计增强后每个类别的样本数量，包括原始数量和新增数量
            enhanced_counts[targets] += len(img_group) + len(maskmix_targets)

        # 打印增强后每个类别的样本数量
        for label, count in enhanced_counts.items():
            print(f"Class {label}: Enhanced sample count: {count}")
        print('Whole train set size:', len(processed_images))
        return processed_images, processed_labels
```

**Comment by fistyee** [comment]
数据集的原始增强操作正常做就可以，如果是项目提点需要你甚至可以添加Randaugmentation, cutout等单视图数据增强。你可以向你代码里这样只对数量少的类别做MaskMix操作，也可以在采样一个batch的数据时候使用平衡采样，增加尾部数据的采样概率，再使用MaskMix数据增强。通常来说MaskMix与其他双视图混合增强方法如，Mixup混合使用效果也会更好

**Comment by fistyee** [comment]
> 您好，我想请问一下
> 
> 1. 就是我如果用了maskmix还需要对数据集做data_transforms的增强操作吗
> 
> ```
> data_transforms = transforms.Compose([
>         transforms.Resize((112, 112)),
>         transforms.RandomHorizontalFlip(),
>         transforms.RandomApply([
>             transforms.RandomRotation(5),
>             transforms.RandomCrop(112, padding=8)
>         ], p=0.2),
>         transforms.ToTensor(),
>         transforms.Normalize(mean=[0.485, 0.456, 0.406],
>                              std=[0.229, 0.224, 0.225]),
>         transforms.RandomErasing(scale=(0.02, 0.25)),
>     ])
>     train_dataset = datasets.ImageFolder(f'{args.raf_path}/train',transform = data_transforms)
>     # 应用处理方法
>     processed_images, processed_labels = apply_maskmix(train_dataset, scale_, scale, label_smoothing, num_classes)
> ```
> 
> 2. 我换了另外一种方式，就是输出了混淆矩阵，对混淆矩阵里面类别数量少，并且准确率低的几个类别做操作，下面是我的代码
> 
> ```
>  # 定义一个方法来处理目标类别为 1、2、4 和 6 的图像
>    def apply_maskmix(train_dataset, scale_, scale, label_smoothing, num_classes):
>        enhanced_counts = defaultdict(int)
>        grouped_imgs = defaultdict(list)
>        processed_images = []
>        processed_labels = []
> 
>        for img, target in train_dataset:
>            if target in [1, 2, 4, 6]:
>                grouped_imgs[target].append((img, target))  # 将图片和标签一起存储
>            else:
>                # 对于不需要增强的类别，直接添加原始样本
>                processed_images.append(img)
>                processed_labels.append(target)
>                enhanced_counts[target] += 1
> 
>        for targets, img_group in grouped_imgs.items():
>           
>            img_group_tensor = torch.stack([(item[0]) for item in img_group])  # 取出图片张量
>            maskmix_imgs, maskmix_targets, _, _ = MaskMix(img_group_tensor, None, alpha, mask_num, scale_,
>                                                          scale, torch.tensor([targets] * len(img_group)),
>                                                          None, label_smoothing, num_classes)
>            processed_images.extend(maskmix_imgs.tolist())
>            processed_labels.extend(maskmix_targets.tolist())
>            # 统计增强后每个类别的样本数量，包括原始数量和新增数量
>            enhanced_counts[targets] += len(img_group) + len(maskmix_targets)
> 
>        # 打印增强后每个类别的样本数量
>        for label, count in enhanced_counts.items():
>            print(f"Class {label}: Enhanced sample count: {count}")
>        print('Whole train set size:', len(processed_images))
>        return processed_images, processed_labels
> ```

数据集的原始增强操作正常做就可以，如果是项目提点需要你甚至可以添加Randaugmentation, cutout等单视图数据增强。你可以向你代码里这样只对数量少的类别做MaskMix操作，也可以在采样一个batch的数据时候使用平衡采样，增加尾部数据的采样概率，再使用MaskMix数据增强。通常来说MaskMix与其他双视图混合增强方法如，Mixup混合使用效果也会更好
